### ISSUE TITLE
[question] mesh scale and coordinate doesn't match to real size using cctag

### ISSUE THREAD

**Issue by stupidyoh** [alicevision/Meshroom#2393]
**Describe the problem**
Hello, thank you for your great work.
I'm trying to reconstruct model using cctag3 following the meshroom wiki
Reconstruction works well, but scaling to real size and orientation doesn't work well...
(For example, the real width of cheeze-it is about 155mm but the reconstructed width is about 500mm)
And also, the distance between the markers doesn't match real value(2*200mm)

Is there any problem with my work or setup?
How can I fix it?

Thank you.

P.S. 
The images was captured by Intel Realsense D435(1920x1080).
Does the absence of metadata affect the result or error?
(And also I want to know how to apply focal length and sensor width manually for D435)

**Screenshots**
![image](https://github.com/alicevision/Meshroom/assets/121854081/38502c3c-e4de-4fb5-a5a6-53f8fd10cc63)
![image](https://github.com/alicevision/Meshroom/assets/121854081/40ba6f16-2f26-468c-a384-d4979636f776)
This is my sample image of target object with cctag(id 0, 1, 2 and 3)
There is 56 images in image gallery.

![image](https://github.com/alicevision/Meshroom/assets/121854081/dc4f61cf-29f4-4db4-9a19-82df9af0abda)
Here is my node connect.

![image](https://github.com/alicevision/Meshroom/assets/121854081/cdbacb3d-2b36-4cbe-ad24-b858502b4a73)
![image](https://github.com/alicevision/Meshroom/assets/121854081/5b79fb14-38b0-42f4-9a01-ebe3c91bc8f2)
![image](https://github.com/alicevision/Meshroom/assets/121854081/bc8f143e-ded0-465b-9275-1acc56ddb380)
Here are the detail setup for each node
I set my parameter in "SfMTransform" node following 
Additional Scale : 2 (because the distance between each marker is 400mm)
marker 0 : X1 Y0 Z1
marker 1 : X-1 Y0 Z1
marker 2 : X-1 Y0 Z-1
marker 3 : X1 Y0 Z-1

![image](https://github.com/alicevision/Meshroom/assets/121854081/d96d38b1-4740-41b3-b517-cadf8c735f98)
I think the result of reconstruction is great!
But the axis is under the mesh of floor, not on the table.

**Log**
![스크린샷 2024-04-29 165558](https://github.com/alicevision/Meshroom/assets/121854081/8f67f0c2-3ec6-45d6-8ffb-1517bf1621b1)
I think the distance should be about 4.0
But the result of SfMDistances is 2.3

**Desktop (please complete the following and other pertinent information):**
 - OS: Ubuntu 20.04
 - Python version : 3.9
 - Qt/PySide version 3.1 
 - Meshroom version: 2023.3.0

**Comment by msanta** [comment]
You forgot to specify the marker IDs in the Transform node.

**Comment by stupidyoh** [comment]
I'm sorry. 
The example picture was not appropriate.
The scale was different from the real one when the ID was given before.

Here is the result with marker IDs

![image](https://github.com/alicevision/Meshroom/assets/121854081/33d3fb20-5ab9-4efd-a092-2797c055c9cc)

![image](https://github.com/alicevision/Meshroom/assets/121854081/50b5f776-e479-4a93-a15d-24a871abc57e)
Still have axis under the mesh

![image](https://github.com/alicevision/Meshroom/assets/121854081/df08b072-1c6b-46c9-91c5-bb53ea01eaeb)
And the result mesh still has 500mm width

**Comment by natowi** [comment]
Your markers are 400mm apart?

To scale the Model to this known size, set the Additional Scale to 4  -->(100*4). 
(The coordinate scale is 100mm per unit)

![grafik](https://github.com/alicevision/Meshroom/assets/8138537/32739068-bb40-4219-8afe-28ba7551dcb8)

**Comment by stupidyoh** [comment]
Yes, the distance between each center of marker is 400mm
And I'm trying various parameter values but...
unfortunately there is no change in the result(500mm width for cheeze-it)

As you say,
scale : 4 / Markers : 0(1, 0, 1), 1(-1, 0, 1), 2(-1, 0, -1), 3(1, 0, -1) -> SfMDistance : 8.0   / Width : 500mm

And
scale : 2 / Markers : 0(1, 0, 1), 1(-1, 0, 1), 2(-1, 0, -1), 3(1, 0, -1) -> SfMDistance : 4.0   / Width : 500mm
scale : 1 / Markers : 0(2, 0, 2), 1(-2, 0, 2), 2(-2, 0, -2), 3(2, 0, -2) -> SfMDistance : 4.0   / Width : 500mm
scale : 4 / Markers : 0(2, 0, 2), 1(-2, 0, 2), 2(-2, 0, -2), 3(2, 0, -2) -> SfMDistance : 16.0 / Width : 500mm
etc...

All of trial give me the same result of mesh... 
(Scale, orientation, coordinates don't match)

![image](https://github.com/alicevision/Meshroom/assets/121854081/5542bafc-48ee-4ae1-956a-006e5143e8f1)
This is my node connect after SfM
Do I need additional connect?
Or should I look elsewhere?

**Comment by msanta** [comment]
What is the result if you set the distance between just two markers?

**Comment by stupidyoh** [comment]
It's about 1.2m
I opened and checked the obj file in two tools, meshlab and blender

![image](https://github.com/alicevision/Meshroom/assets/121854081/9595bb57-6cc7-43e9-91ff-8ee1652abd9c)

**Comment by stupidyoh** [comment]
And the result is also same when activate just two markers

![image](https://github.com/alicevision/Meshroom/assets/121854081/74e4d1c0-1f41-42ba-8abe-b904d54b060f)
![image](https://github.com/alicevision/Meshroom/assets/121854081/76c6f177-63f9-4a1e-a3f1-85a308f9e864)
![image](https://github.com/alicevision/Meshroom/assets/121854081/38c7dbf9-b0b6-4acc-94ab-86654cad787a)

**Comment by msanta** [comment]
I believe markers 0 and 2 are diagonal from each other. Might make life easier to position adjacent markers. In any case the transform worked as 0 <=> 2 is 5.78.

Silly question, are you getting the model file from the SfMTransform folder? If so you might want to try exporting just the CCTag features using PLY format and measuring the distances of the four points. They should match what Meshroom is reporting.

**Comment by stupidyoh** [comment]
I have checked my model from "Meshing" folder (click the node and open folder)

My goal is to reconstruct textured mesh file with real-world scale and apply it other field.
After clear this cctag problem, I plan to extract target object mesh using custom bounding box in "Meshing" node.
So the scaled and oriented mesh that fits my intentions is needed...

**Comment by stupidyoh** [comment]
Is the "meshing" node where I can get the mesh I want?
If not, where can I get the scaled and orientated mesh?

**Comment by msanta** [comment]
Yes, your node setup is correct. The meshing process has used the transformed SfMData.

**Comment by stupidyoh** [comment]
I tried to delete all my cache and reboot meshroom, then run whole progress once again.
I think "SfMTransform" works well but transformed data isn't applied properly on the "meshing" node

![image](https://github.com/alicevision/Meshroom/assets/121854081/0974e655-d60e-40d2-a282-5f420074dd3f)
Result of SfM(Upper) and SfMTransform(Lower)
SfMTransform is upside down

![image](https://github.com/alicevision/Meshroom/assets/121854081/a804dc5f-d83c-4994-91c9-8f2e8dc0b92a)
Reconstructed mesh matches SfM, not SfMTransform

![image](https://github.com/alicevision/Meshroom/assets/121854081/9c8ee334-bed6-4781-add0-17bbaba15acf)
Does CCTag textured on the floor affect these error?
(The CCTag is only on the plate, not on the bottom floor)

**Comment by msanta** [comment]
Here you can see the point cloud from the SfM node and the mesh which has been oriented via the SfMTransform node.

![image](https://github.com/alicevision/Meshroom/assets/5391990/b9fb2fed-71c9-4988-b147-5c65f26c134e)

SfM -> SfMTransform -> PrepareDepthScene -> DepthMap -> DepthMapFilter -> Meshing

Or do it this way:
![image](https://github.com/alicevision/Meshroom/assets/5391990/fc683f3b-97e1-43ab-9420-68f41befe6ce)

**Comment by stupidyoh** [comment]
It works well for the first node setup!
Thank you for your considerable feedback!!!
